File: akka-cluster-sharding-typed/src/test/scala/docs/delivery/ShardingDocExample.scala
/*
 * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.delivery

//#imports
import scala.concurrent.Future
import scala.concurrent.duration._
import scala.util.Failure
import scala.util.Success

import akka.Done
import akka.actor.typed.ActorRef
import akka.actor.typed.Behavior
import akka.actor.typed.delivery.ConsumerController
import akka.actor.typed.scaladsl.ActorContext
import akka.actor.typed.scaladsl.Behaviors
import akka.cluster.sharding.typed.delivery.ShardingConsumerController
import akka.util.Timeout

//#imports

object ShardingDocExample {

  //#consumer
  trait DB {
    def save(id: String, value: TodoList.State): Future[Done]
    def load(id: String): Future[TodoList.State]
  }

  object TodoList {

    sealed trait Command

    final case class AddTask(item: String) extends Command
    final case class CompleteTask(item: String) extends Command

    private final case class InitialState(state: State) extends Command
    private final case class SaveSuccess(confirmTo: ActorRef[ConsumerController.Confirmed]) extends Command
    private final case class DBError(cause: Throwable) extends Command

    private final case class CommandDelivery(command: Command, confirmTo: ActorRef[ConsumerController.Confirmed])
        extends Command

    final case class State(tasks: Vector[String])

    def apply(
        id: String,
        db: DB,
        consumerController: ActorRef[ConsumerController.Start[Command]]): Behavior[Command] = {
      Behaviors.setup[Command] { context =>
        new TodoList(context, id, db).start(consumerController)
      }
    }

  }

  class TodoList(context: ActorContext[TodoList.Command], id: String, db: DB) {
    import TodoList._

    private def start(consumerController: ActorRef[ConsumerController.Start[Command]]): Behavior[Command] = {
      context.pipeToSelf(db.load(id)) {
        case Success(value) => InitialState(value)
        case Failure(cause) => DBError(cause)
      }

      Behaviors.receiveMessagePartial {
        case InitialState(state) =>
          val deliveryAdapter: ActorRef[ConsumerController.Delivery[Command]] = context.messageAdapter { delivery =>
            CommandDelivery(delivery.message, delivery.confirmTo)
          }
          consumerController ! ConsumerController.Start(deliveryAdapter)
          active(state)
        case DBError(cause) =>
          throw cause
      }
    }

    private def active(state: State): Behavior[Command] = {
      Behaviors.receiveMessagePartial {
        case CommandDelivery(AddTask(item), confirmTo) =>
          val newState = state.copy(tasks = state.tasks :+ item)
          save(newState, confirmTo)
          active(newState)
        case CommandDelivery(CompleteTask(item), confirmTo) =>
          val newState = state.copy(tasks = state.tasks.filterNot(_ == item))
          save(newState, confirmTo)
          active(newState)
        case SaveSuccess(confirmTo) =>
          confirmTo ! ConsumerController.Confirmed
          Behaviors.same
        case DBError(cause) =>
          throw cause
      }
    }

    private def save(newState: State, confirmTo: ActorRef[ConsumerController.Confirmed]): Unit = {
      context.pipeToSelf(db.save(id, newState)) {
        case Success(_)     => SaveSuccess(confirmTo)
        case Failure(cause) => DBError(cause)
      }
    }
  }
  //#consumer

  //#producer
  import akka.cluster.sharding.typed.delivery.ShardingProducerController

  object TodoService {
    sealed trait Command

    final case class UpdateTodo(listId: String, item: String, completed: Boolean, replyTo: ActorRef[Response])
        extends Command

    sealed trait Response
    case object Accepted extends Response
    case object Rejected extends Response
    case object MaybeAccepted extends Response

    private final case class WrappedRequestNext(requestNext: ShardingProducerController.RequestNext[TodoList.Command])
        extends Command
    private final case class Confirmed(originalReplyTo: ActorRef[Response]) extends Command
    private final case class TimedOut(originalReplyTo: ActorRef[Response]) extends Command

    def apply(producerController: ActorRef[ShardingProducerController.Command[TodoList.Command]]): Behavior[Command] = {
      Behaviors.setup { context =>
        new TodoService(context).start(producerController)
      }
    }

  }

  class TodoService(context: ActorContext[TodoService.Command]) {
    import TodoService._

    private implicit val askTimeout: Timeout = 5.seconds

    private def start(
        producerController: ActorRef[ShardingProducerController.Start[TodoList.Command]]): Behavior[Command] = {
      val requestNextAdapter: ActorRef[ShardingProducerController.RequestNext[TodoList.Command]] =
        context.messageAdapter(WrappedRequestNext.apply)
      producerController ! ShardingProducerController.Start(requestNextAdapter)

      Behaviors.receiveMessagePartial {
        case WrappedRequestNext(next) =>
          active(next)
        case UpdateTodo(_, _, _, replyTo) =>
          // not hooked up with shardingProducerController yet
          replyTo ! Rejected
          Behaviors.same
      }
    }

    private def active(requestNext: ShardingProducerController.RequestNext[TodoList.Command]): Behavior[Command] = {
      Behaviors.receiveMessage {
        case WrappedRequestNext(next) =>
          active(next)

        case UpdateTodo(listId, item, completed, replyTo) =>
          if (requestNext.bufferedForEntitiesWithoutDemand.getOrElse(listId, 0) >= 100)
            replyTo ! Rejected
          else {
            val requestMsg = if (completed) TodoList.CompleteTask(item) else TodoList.AddTask(item)
            context.ask[ShardingProducerController.MessageWithConfirmation[TodoList.Command], Done](
              requestNext.askNextTo,
              askReplyTo => ShardingProducerController.MessageWithConfirmation(listId, requestMsg, askReplyTo)) {
              case Success(Done) => Confirmed(replyTo)
              case Failure(_)    => TimedOut(replyTo)
            }
          }
          Behaviors.same

        case Confirmed(originalReplyTo) =>
          originalReplyTo ! Accepted
          Behaviors.same

        case TimedOut(originalReplyTo) =>
          originalReplyTo ! MaybeAccepted
          Behaviors.same
      }
    }

  }
  //#producer

  def illustrateInit(): Unit = {
    Behaviors.setup[Nothing] { context =>
      //#init
      import akka.cluster.sharding.typed.scaladsl.ClusterSharding
      import akka.cluster.sharding.typed.scaladsl.Entity
      import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
      import akka.cluster.typed.Cluster

      val db: DB = ???

      val system = context.system

      val TypeKey = EntityTypeKey[ConsumerController.SequencedMessage[TodoList.Command]]("todo")

      val region = ClusterSharding(system).init(Entity(TypeKey)(entityContext =>
        ShardingConsumerController(start => TodoList(entityContext.entityId, db, start))))

      val selfAddress = Cluster(system).selfMember.address
      val producerId = s"todo-producer-${selfAddress.host}:${selfAddress.port}"

      val producerController =
        context.spawn(ShardingProducerController(producerId, region, durableQueueBehavior = None), "producerController")

      context.spawn(TodoService(producerController), "producer")
      //#init

      Behaviors.empty
    }
  }

}

File: akka-cluster-sharding-typed/src/test/java/jdocs/delivery/ShardingDocExample.java
/*
 * Copyright (C) 2020-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.delivery;

// #imports
import akka.Done;
import akka.actor.Address;
import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.delivery.ConsumerController;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;

import java.time.Duration;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletionStage;

// #imports

// #producer
import akka.cluster.sharding.typed.delivery.ShardingProducerController;

// #producer

// #init
import akka.cluster.sharding.typed.delivery.ShardingConsumerController;
import akka.cluster.sharding.typed.ShardingEnvelope;
import akka.cluster.sharding.typed.javadsl.ClusterSharding;
import akka.cluster.sharding.typed.javadsl.Entity;
import akka.cluster.sharding.typed.javadsl.EntityTypeKey;
import akka.cluster.typed.Cluster;
import akka.actor.typed.ActorSystem;

// #init

interface ShardingDocExample {

  // #consumer

  interface DB {
    CompletionStage<Done> save(String id, TodoList.State state);

    CompletionStage<TodoList.State> load(String id);
  }

  public class TodoList {
    interface Command {}

    public static class AddTask implements Command {
      public final String item;

      public AddTask(String item) {
        this.item = item;
      }
    }

    public static class CompleteTask implements Command {
      public final String item;

      public CompleteTask(String item) {
        this.item = item;
      }
    }

    private static class InitialState implements Command {
      final State state;

      private InitialState(State state) {
        this.state = state;
      }
    }

    private static class SaveSuccess implements Command {
      final ActorRef<ConsumerController.Confirmed> confirmTo;

      private SaveSuccess(ActorRef<ConsumerController.Confirmed> confirmTo) {
        this.confirmTo = confirmTo;
      }
    }

    private static class DBError implements Command {
      final Exception cause;

      private DBError(Throwable cause) {
        if (cause instanceof Exception) this.cause = (Exception) cause;
        else this.cause = new RuntimeException(cause.getMessage(), cause);
      }
    }

    private static class CommandDelivery implements Command {
      final Command command;
      final ActorRef<ConsumerController.Confirmed> confirmTo;

      private CommandDelivery(Command command, ActorRef<ConsumerController.Confirmed> confirmTo) {
        this.command = command;
        this.confirmTo = confirmTo;
      }
    }

    public static class State {
      public final List<String> tasks;

      public State(List<String> tasks) {
        this.tasks = Collections.unmodifiableList(tasks);
      }

      public State add(String task) {
        ArrayList<String> copy = new ArrayList<>(tasks);
        copy.add(task);
        return new State(copy);
      }

      public State remove(String task) {
        ArrayList<String> copy = new ArrayList<>(tasks);
        copy.remove(task);
        return new State(copy);
      }
    }

    public static Behavior<Command> create(
        String id, DB db, ActorRef<ConsumerController.Start<Command>> consumerController) {
      return Init.create(id, db, consumerController);
    }

    private static Behavior<Command> onDBError(DBError error) throws Exception {
      throw error.cause;
    }

    static class Init extends AbstractBehavior<Command> {

      private final String id;
      private final DB db;
      private final ActorRef<ConsumerController.Start<Command>> consumerController;

      private Init(
          ActorContext<Command> context,
          String id,
          DB db,
          ActorRef<ConsumerController.Start<Command>> consumerController) {
        super(context);
        this.id = id;
        this.db = db;
        this.consumerController = consumerController;
      }

      static Behavior<Command> create(
          String id, DB db, ActorRef<ConsumerController.Start<Command>> consumerController) {
        return Behaviors.setup(
            context -> {
              context.pipeToSelf(
                  db.load(id),
                  (state, exc) -> {
                    if (exc == null) return new InitialState(state);
                    else return new DBError(exc);
                  });

              return new Init(context, id, db, consumerController);
            });
      }

      @Override
      public Receive<Command> createReceive() {
        return newReceiveBuilder()
            .onMessage(InitialState.class, this::onInitialState)
            .onMessage(DBError.class, TodoList::onDBError)
            .build();
      }

      private Behavior<Command> onInitialState(InitialState initial) {
        ActorRef<ConsumerController.Delivery<Command>> deliveryAdapter =
            getContext()
                .messageAdapter(
                    ConsumerController.deliveryClass(),
                    d -> new CommandDelivery(d.message(), d.confirmTo()));
        consumerController.tell(new ConsumerController.Start<>(deliveryAdapter));

        return Active.create(id, db, initial.state);
      }
    }

    static class Active extends AbstractBehavior<Command> {

      private final String id;
      private final DB db;
      private State state;

      private Active(ActorContext<Command> context, String id, DB db, State state) {
        super(context);
        this.id = id;
        this.db = db;
        this.state = state;
      }

      static Behavior<Command> create(String id, DB db, State state) {
        return Behaviors.setup(context -> new Active(context, id, db, state));
      }

      @Override
      public Receive<Command> createReceive() {
        return newReceiveBuilder()
            .onMessage(CommandDelivery.class, this::onDelivery)
            .onMessage(SaveSuccess.class, this::onSaveSuccess)
            .onMessage(DBError.class, TodoList::onDBError)
            .build();
      }

      private Behavior<Command> onDelivery(CommandDelivery delivery) {
        if (delivery.command instanceof AddTask) {
          AddTask addTask = (AddTask) delivery.command;
          state = state.add(addTask.item);
          save(state, delivery.confirmTo);
          return this;
        } else if (delivery.command instanceof CompleteTask) {
          CompleteTask completeTask = (CompleteTask) delivery.command;
          state = state.remove(completeTask.item);
          save(state, delivery.confirmTo);
          return this;
        } else {
          return Behaviors.unhandled();
        }
      }

      private void save(State newState, ActorRef<ConsumerController.Confirmed> confirmTo) {
        getContext()
            .pipeToSelf(
                db.save(id, newState),
                (state, exc) -> {
                  if (exc == null) return new SaveSuccess(confirmTo);
                  else return new DBError(exc);
                });
      }

      private Behavior<Command> onSaveSuccess(SaveSuccess success) {
        success.confirmTo.tell(ConsumerController.confirmed());
        return this;
      }
    }
  }
  // #consumer

  // #producer
  public class TodoService {

    interface Command {}

    public static class UpdateTodo implements Command {
      public final String listId;
      public final String item;
      public final boolean completed;
      public final ActorRef<Response> replyTo;

      public UpdateTodo(String listId, String item, boolean completed, ActorRef<Response> replyTo) {
        this.listId = listId;
        this.item = item;
        this.completed = completed;
        this.replyTo = replyTo;
      }
    }

    public enum Response {
      ACCEPTED,
      REJECTED,
      MAYBE_ACCEPTED
    }

    private static class Confirmed implements Command {
      final ActorRef<Response> originalReplyTo;

      private Confirmed(ActorRef<Response> originalReplyTo) {
        this.originalReplyTo = originalReplyTo;
      }
    }

    private static class TimedOut implements Command {
      final ActorRef<Response> originalReplyTo;

      private TimedOut(ActorRef<Response> originalReplyTo) {
        this.originalReplyTo = originalReplyTo;
      }
    }

    private static class WrappedRequestNext implements Command {
      final ShardingProducerController.RequestNext<TodoList.Command> next;

      private WrappedRequestNext(ShardingProducerController.RequestNext<TodoList.Command> next) {
        this.next = next;
      }
    }

    public static Behavior<Command> create(
        ActorRef<ShardingProducerController.Command<TodoList.Command>> producerController) {
      return Init.create(producerController);
    }

    static class Init extends AbstractBehavior<TodoService.Command> {

      static Behavior<Command> create(
          ActorRef<ShardingProducerController.Command<TodoList.Command>> producerController) {
        return Behaviors.setup(
            context -> {
              ActorRef<ShardingProducerController.RequestNext<TodoList.Command>>
                  requestNextAdapter =
                      context.messageAdapter(
                          ShardingProducerController.requestNextClass(), WrappedRequestNext::new);
              producerController.tell(new ShardingProducerController.Start<>(requestNextAdapter));

              return new Init(context);
            });
      }

      private Init(ActorContext<Command> context) {
        super(context);
      }

      @Override
      public Receive<Command> createReceive() {
        return newReceiveBuilder()
            .onMessage(WrappedRequestNext.class, w -> Active.create(w.next))
            .onMessage(
                UpdateTodo.class,
                command -> {
                  // not hooked up with shardingProducerController yet
                  command.replyTo.tell(Response.REJECTED);
                  return this;
                })
            .build();
      }
    }

    static class Active extends AbstractBehavior<TodoService.Command> {

      private ShardingProducerController.RequestNext<TodoList.Command> requestNext;

      static Behavior<Command> create(
          ShardingProducerController.RequestNext<TodoList.Command> requestNext) {
        return Behaviors.setup(context -> new Active(context, requestNext));
      }

      private Active(
          ActorContext<Command> context,
          ShardingProducerController.RequestNext<TodoList.Command> requestNext) {
        super(context);
        this.requestNext = requestNext;
      }

      @Override
      public Receive<Command> createReceive() {
        return newReceiveBuilder()
            .onMessage(WrappedRequestNext.class, this::onRequestNext)
            .onMessage(UpdateTodo.class, this::onUpdateTodo)
            .onMessage(Confirmed.class, this::onConfirmed)
            .onMessage(TimedOut.class, this::onTimedOut)
            .build();
      }

      private Behavior<Command> onRequestNext(WrappedRequestNext w) {
        requestNext = w.next;
        return this;
      }

      private Behavior<Command> onUpdateTodo(UpdateTodo command) {
        Integer buffered = requestNext.getBufferedForEntitiesWithoutDemand().get(command.listId);
        if (buffered != null && buffered >= 100) {
          command.replyTo.tell(Response.REJECTED);
        } else {
          TodoList.Command requestMsg;
          if (command.completed) requestMsg = new TodoList.CompleteTask(command.item);
          else requestMsg = new TodoList.AddTask(command.item);
          getContext()
              .ask(
                  Done.class,
                  requestNext.askNextTo(),
                  Duration.ofSeconds(5),
                  askReplyTo ->
                      new ShardingProducerController.MessageWithConfirmation<>(
                          command.listId, requestMsg, askReplyTo),
                  (done, exc) -> {
                    if (exc == null) return new Confirmed(command.replyTo);
                    else return new TimedOut(command.replyTo);
                  });
        }
        return this;
      }

      private Behavior<Command> onConfirmed(Confirmed confirmed) {
        confirmed.originalReplyTo.tell(Response.ACCEPTED);
        return this;
      }

      private Behavior<Command> onTimedOut(TimedOut timedOut) {
        timedOut.originalReplyTo.tell(Response.MAYBE_ACCEPTED);
        return this;
      }
    }
  }
  // #producer

  static void illustrateInit() {
    Behaviors.setup(
        context -> {
          // #init
          final DB db = theDatabaseImplementation();

          ActorSystem<Void> system = context.getSystem();

          EntityTypeKey<ConsumerController.SequencedMessage<TodoList.Command>> entityTypeKey =
              EntityTypeKey.create(ShardingConsumerController.entityTypeKeyClass(), "todo");

          ActorRef<ShardingEnvelope<ConsumerController.SequencedMessage<TodoList.Command>>> region =
              ClusterSharding.get(system)
                  .init(
                      Entity.of(
                          entityTypeKey,
                          entityContext ->
                              ShardingConsumerController.create(
                                  start ->
                                      TodoList.create(entityContext.getEntityId(), db, start))));

          Address selfAddress = Cluster.get(system).selfMember().address();
          String producerId = "todo-producer-" + selfAddress.hostPort();

          ActorRef<ShardingProducerController.Command<TodoList.Command>> producerController =
              context.spawn(
                  ShardingProducerController.create(
                      TodoList.Command.class, producerId, region, Optional.empty()),
                  "producerController");

          context.spawn(TodoService.create(producerController), "producer");
          // #init

          return Behaviors.empty();
        });
  }

  static DB theDatabaseImplementation() {
    return null;
  }
}
